### PR TITLE
try push to dev branch v.1

### PR DESCRIPTION
**Pull request by Xabaraz**
code has excess if()

### CODE REVIEW

**Comment by aesthetic77** [comment]
you need to create .gitignore file, so you will exclude unnecessary files to be comitted and pushed (for ex. .idea folder)

File: src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        todo.setUserName(userName);
+        todo.setUserTime(0L);
+        todo.setLastVisit(new Date());
+        todo.save();
+    }
+
+    @Override
+    public Todo findTodoByName(String userName) {
+        List<Todo> todo = newArrayList(ao.find(Todo.class, Query.select().where("USER_NAME = ?", userName)));
+        if (todo.isEmpty()) return null;
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java]: you can use Java8 stream api for such cases (ex.  .stream().findFirst().orElse(null))

File: src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        return todo.get(0);
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java]: you can use Java8 stream api for such cases (ex. array.stream().findFirst().orElse(null))

File: src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    public void setUserTime(String userName, Long userTime) {
+        List<Todo> todo = newArrayList(ao.find(Todo.class, Query.select().where("USER_NAME = ?", userName)));
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java]: do not be repeating, you can reuse findTodoByName function

File: src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        todo.get(0).setUserTime(userTime);
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java]: missed null checks

File: src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import net.java.ao.Query;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.Date;
+import java.util.List;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+@Scanned
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java]: not sure, but @Scanned or @Named may be redundant here, please recheck

File: src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public void setUserTime(String userName, Long userTime) {
+        List<Todo> todo = newArrayList(ao.find(Todo.class, Query.select().where("USER_NAME = ?", userName)));
+        todo.get(0).setUserTime(userTime);
+        todo.get(0).save();
+    }
+
+    @Override
+    public Date getLastVisit(String userName) {
+        List<Todo> todo = newArrayList(ao.find(Todo.class, Query.select().where("USER_NAME = ?", userName)));
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java]: reuse

File: src/main/java/com/atlassian/tutorial/rest/MyRestResource.java [hunk 7]
```diff
@@ -1,41 +1,53 @@
...
+        Date timeNow = new Date();
 
-    private String getMessageFromKey(String key) {
-        // In reality, this data would come from a database or some component
-        // within the hosting application, for demonstration purpopses I will
-        // just return the key
-        return key;
+        Todo todo = aOtodo.findTodoByName(userName);
+        if (todo == null) {
+            aOtodo.createTodo(userName, timeNow);
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/rest/MyRestResource.java]: you can make createTodo function return created Todo, so you won't have to search it later with findTodoByName

File: src/main/java/com/atlassian/tutorial/rest/MyRestResource.java [hunk 8]
```diff
@@ -1,41 +1,53 @@
...
+            todo = aOtodo.findTodoByName(userName);
+        }
+        if (todo.getLastVisit().getTime() / 1000 != timeNow.getTime() / 1000) {
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/rest/MyRestResource.java]: you don't need to divide by 1000 in both sides of comparison (useless)

File: src/main/java/com/atlassian/tutorial/rest/MyRestResource.java [hunk 9]
```diff
@@ -1,41 +1,53 @@
...
+            Long userCurentTime = todo.getUserTime();
+            Long pause = (timeNow.getTime() / 1000) - (todo.getLastVisit().getTime() / 1000);
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/rest/MyRestResource.java]: you can divide by 1000 after substitution

File: src/main/java/com/atlassian/tutorial/rest/MyRestResourceModel.java [hunk 10]
```diff
@@ -1,40 +1,48 @@
...
+    private String userTime;
+    private String lastVisit;
+
 
     public MyRestResourceModel() {
     }
 
-    public MyRestResourceModel(String message) {
-        this.message = message;
+    public MyRestResourceModel(String userName, String userTime,String lastVisit) {
```
Comment (aesthetic77) [inline src/main/java/com/atlassian/tutorial/rest/MyRestResourceModel.java]: always remember to format code (Ctrl+Shift+L) before commiting

File: src/main/resources/js/myPluginTST.js [hunk 11]
```diff
@@ -1,2 +1,34 @@
...
+                },
+                error: function (error) {
+                    console.log(error)
+                }
+            });
+        }
+        , 1000);
+};
+
+function Clock(data) {
```
Comment (aesthetic77) [inline src/main/resources/js/myPluginTST.js]: use code convention for javascript (first lowercase symbol in function names)

File: src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.Date;
+import java.util.List;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+@Scanned
+@Named
+public class AOtodo implements TodoService {
```
Comment (EgorBrusnev) [inline src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java]: use camlelcase in names, such as AOTodo or TodoAO

File: src/main/java/com/atlassian/tutorial/rest/MyRestResource.java [hunk 13]
```diff
@@ -1,41 +1,53 @@
...
-    @GET
-    @AnonymousAllowed
-    @Produces({MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML})
-    public Response getMessage(@QueryParam("key") String key)
-    {
-        if(key!=null)
-            return Response.ok(new MyRestResourceModel(key, getMessageFromKey(key))).build();
-        else
-            return Response.ok(new MyRestResourceModel("default","Hello World")).build();
+    private AOtodo aOtodo;
```
Comment (EgorBrusnev) [inline src/main/java/com/atlassian/tutorial/rest/MyRestResource.java]: the same, name variable aoTodo or todoAo

File: src/main/java/com/atlassian/tutorial/ao/todo/Todo.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
+package com.atlassian.tutorial.ao.todo;
+
+import net.java.ao.Entity;
+import net.java.ao.Preload;
+
+import java.util.Date;
+
+@Preload
+public interface Todo extends Entity {
```
Comment (EgorBrusnev) [inline src/main/java/com/atlassian/tutorial/ao/todo/Todo.java]: you need to use meaningful names ( this entity is for user's time spent on the site, not todo)

File: src/main/resources/js/myPluginTST.js [hunk 15]
```diff
@@ -1,2 +1,34 @@
...
+        function () {
+            jQuery.ajax({
+                type: "get",
+                url: "http://localhost:2990/jira/rest/myrestresource/1.0/message",
+                success: function (data) {
+                    console.log(data);
+                    Clock(data);
+                },
+                error: function (error) {
+                    console.log(error)
```
Comment (EgorBrusnev) [inline src/main/resources/js/myPluginTST.js]: add error handling, for example show message in UI or something else

File: src/main/resources/js/myPluginTST.js [hunk 16]
```diff
@@ -1,2 +1,34 @@
...
+        , 1000);
+};
+
+function Clock(data) {
+    var seconds = data.userTime;
+    var minutes = (seconds / 60);
+    if (minutes > 60) {
+        minutes = minutes % 60
+    };
+    if(1) {};
```
Comment (EgorBrusnev) [inline src/main/resources/js/myPluginTST.js]: what is this?

File: src/main/resources/js/myPluginTST.js [hunk 17]
```diff
@@ -1,2 +1,34 @@
...
+    if (minutes < 1) minutes = 0;
+    var hours = (seconds / 3600);
+    hours = hours - (hours % 1);
+    if (hours < 1) hours = 0;
+    console.log(seconds);
+    document.getElementById("hours").innerHTML = hours - (hours % 1);
```
Comment (EgorBrusnev) [inline src/main/resources/js/myPluginTST.js]: use jQuery

File: myPluginTST.iml [hunk 18]
```diff
@@ -1,5 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (EgorBrusnev) [inline myPluginTST.iml]: add this file to ``.gitignore``

File: pom.xml [hunk 19]
```diff
@@ -116,6 +116,33 @@
             <version>2.6.0</version>
             <scope>provided</scope>
         </dependency>
+        <!--<dependency>-->
```
Comment (EgorBrusnev) [inline pom.xml]: remove commented dependencies

File: src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java (hunk 5 above)
Comment (EgorBrusnev) [reply src/main/java/com/atlassian/tutorial/ao/todo/AOtodo.java]: use only ``@Named``

File: src/main/java/com/atlassian/tutorial/rest/MyRestResourceModel.java [hunk 20]
```diff
@@ -1,40 +1,48 @@
...
 
-import javax.xml.bind.annotation.*;
-@XmlRootElement(name = "message")
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlRootElement;
+
+@XmlRootElement(name = "userTime")
 @XmlAccessorType(XmlAccessType.FIELD)
 public class MyRestResourceModel {
```
Comment (EgorBrusnev) [inline src/main/java/com/atlassian/tutorial/rest/MyRestResourceModel.java]: also, don't use My.... class names. Use meaningful names

**Pull request merged**